slides/04-understanding_ownership.qmd

```diff
@@ -2,30 +2,204 @@
 engine: knitr
 title: "4. Understanding Ownership"
 ---
+# Learning objectives 
 
-# Learning objectives
+## Learning objectives 
 
-::: nonincremental
-- THESE ARE NICE TO HAVE BUT NOT ABSOLUTELY NECESSARY
-:::
+1. Understand Rust's ownership system
+2. Work with the borrow checker 
+3. Implement the right ownership model at the right time
 
-::: notes
-- You can add notes on each slide with blocks like this!
-- Load a deck in the browser and type "s" to see these notes.
-:::
+# What Is Ownership?
 
-# SLIDE SECTION
+## Moves
 
-## SLIDE
+```rust
+let x = String::from("hello");
+let x2 = x;
+println!("{x}");
+```
 
-- DENOTE MAJOR SECTIONS WITH `# TITLE` (eg `# Installation`)
-- ADD INDIVIDUAL SLIDES WITH `##` (eg `## rustup on Linux/macOS`)
-- KEEP THEM RELATIVELY SLIDE-LIKE; THESE ARE NOTES, NOT THE BOOK ITSELF.
+## Moves
 
-## SLIDE
+```rust
+let x = String::from("hello");
+take_ownership(x);
+println!("{x}");
 
-# SLIDE SECTION
+fn take_ownership(_name: String) {
+  ()
+}
+```
 
-## SLIDE
+## Moves
 
-## SLIDE
+```rust
+let x = String::from("hello");
+take_ownership(x.clone());
+println!("{x}");
+
+fn take_ownership(_name: String) {
+  ()
+}
+```
+
+## Moves
+
+```{r}
+#| eval: true
+#| echo: false
+knitr::include_graphics("img/clone_memory.png")
+```
+
+## Stack vs Heap
+
+```rust
+let x = 1;
+println!("{x}");
+println!("{x}");
+```
+
+## Stack
+
+```{r}
+#| eval: true
+#| echo: false
+knitr::include_graphics("~/core/active/reference/bookclub-rust/slides/img/stack.png")
```

Comment:
This probably should be just "img/stack.png".   This is why the build is failing I believe.

Reply:
If you put this in as a code suggestion, I can accept it on my phone.

Reply:
```suggestion
knitr::include_graphics("img/stack.png")
```